include/envoy/stream_info/filter_state.h

```diff
@@ -55,6 +56,69 @@ class FilterState {
             (dynamic_cast<const T*>(getDataGeneric(data_name)) != nullptr));
   }
 
+  /**
+   * The addToList, hasListWithName, forEachListItem are variants of the above
+   * functions, that operate on list data. Multiple elements could be added
+   * the list under the same data_name.
+   * @param data_name the name of the list data being set.
+   * @param data an owning pointer to the data to be appended to the list.
+   * Note that data_names for list elements do not share the same namespace as
+   * the data_names for singleton data objects added through setData. All items
+   * added to the list must be of the same type.
+   */
+  template <typename T>
+  void addToList(absl::string_view data_name, std::unique_ptr<Object>&& data) {
+    const auto* list = getList(data_name);
+    if (list != nullptr) {
+      // Check type of first element in the list
+      const T* cast = dynamic_cast<const T*>(list->at(0).get());
+      if (!cast) {
+        throw EnvoyException(
+            fmt::format("List {} does not conform to the specified type", data_name));
+      }
+    }
+
+    addToListGeneric(data_name, std::move(data));
+  }
+
+  /**
+   * @param data_name the name of the list being probed.
+   * @return Whether a list of the type and name specified exists in the
+   * data store.
+   */
+  template <typename T> bool hasList(absl::string_view data_name) const {
+    const auto* list = getList(data_name);
+    return ((list != nullptr) && (dynamic_cast<const T*>(list->at(0).get()) != nullptr));
```

Comment:
I'd leave out the test against dynamic casting (which means that this doesn't need to be a templated function) or actually go back to throwing from this location.  Taking @ggreenway 's point that every code location that accesses a particular test will be using the same type if it's correct code, we'd like to have the throw as early as possible.  And if there's a type mismatch in the current code, you'll end up taking a branch in the consumer that assumes that the data isn't present as opposed to there being a coding error.

If there's a use case that requires a catch in consumer code for this case I'm not yet understanding it--if every code location touching the same name should use the same type, and you only throw when a different type is used, why do you need to catch?

Reply:
I was simply following what you did earlier with hasData<>() and hasDataWithName(). I can remove hasData<> and hasList<> for consistency, if thats what you are implying.